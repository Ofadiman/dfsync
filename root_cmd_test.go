package main

import (
	"bytes"
	"os"
	"path/filepath"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/spf13/cobra"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/suite"
)

type RootCommandSuite struct {
	suite.Suite
	logs              *bytes.Buffer
	command           *cobra.Command
	fileSystemBuilder *FileSystemBuilder
	home              string
}

func (suite *RootCommandSuite) BeforeTest() {
	suite.T().Setenv("LOG_LEVEL", "debug")

	home, _ := os.UserHomeDir()
	suite.home = home

	suite.fileSystemBuilder = &FileSystemBuilder{}

	logsBuffer := new(bytes.Buffer)
	logger := createLogger(logsBuffer)

	suite.logs = logsBuffer
	suite.command = createRootCommand(logger)
	suite.command.SetOut(suite.logs)
	suite.command.SetErr(suite.logs)
}

func (suite *RootCommandSuite) TestShouldExitEarlyWhenConflictResolutionFlagIsInvalid() {
	suite.BeforeTest()

	suite.command.SetArgs([]string{"--conflict-resolution", "bar"})
	suite.command.Execute()

	snaps.MatchSnapshot(suite.T(), suite.logs.String())
}

func (suite *RootCommandSuite) TestShouldExitEarlyWhenPathPassedToSourceDirectoryOptionDoesNotExist() {
	suite.BeforeTest()

	suite.command.SetArgs([]string{"--source-directory", "/foo/bar/buzz"})
	suite.command.Execute()

	snaps.MatchSnapshot(suite.T(), suite.logs.String())
}

func (suite *RootCommandSuite) TestShouldExitEarlyWhenPathPassedToSourceDirectoryOptionIsNotDirectory() {
	suite.BeforeTest()

	filePath := "/tmp/foo.txt"
	cleanup := suite.fileSystemBuilder.File(filePath).Build()
	defer cleanup()

	suite.command.SetArgs([]string{"--source-directory", filePath})
	suite.command.Execute()

	snaps.MatchSnapshot(suite.T(), suite.logs.String())
}

func (suite *RootCommandSuite) TestShouldExitEarlyWhenSourceDirectoryIsEmpty() {
	suite.BeforeTest()

	emptyDirPath := "/tmp/source/"
	cleanup := suite.fileSystemBuilder.Directory(emptyDirPath).Build()
	defer cleanup()

	suite.command.SetArgs([]string{"--source-directory", emptyDirPath})
	suite.command.Execute()

	snaps.MatchSnapshot(suite.T(), suite.logs.String())
}

func (suite *RootCommandSuite) TestShouldDoNothingWhenFileExistsAndNoActionIsPassedToConflictResolutionOption() {
	suite.BeforeTest()

	sourceDirectoryPath := "/tmp/source/"
	sourceFilePath := "/tmp/source/foo.txt"
	homeFilePath := filepath.Join(suite.home, "foo.txt")
	cleanup := suite.fileSystemBuilder.Directory(sourceDirectoryPath).File(sourceFilePath).File(homeFilePath).Build()
	defer cleanup()

	suite.command.SetArgs([]string{"--source-directory", sourceDirectoryPath})
	suite.command.Execute()

	snaps.MatchSnapshot(suite.T(), suite.logs.String())
	homeFileStat, _ := os.Lstat(homeFilePath)
	assert.NotEqual(suite.T(), os.ModeSymlink, homeFileStat.Mode()&os.ModeSymlink)
}

func (suite *RootCommandSuite) TestShouldCreateBackupFileWhenFileExistsAndBackupIsPassedToConflictResolutionOption() {
	suite.BeforeTest()

	sourceDirectoryPath := "/tmp/source/"
	sourceFilePath := "/tmp/source/foo.txt"
	homeFilePath := filepath.Join(suite.home, "foo.txt")
	homeBackupFilePath := homeFilePath + ".bak"
	cleanup := suite.fileSystemBuilder.Directory(sourceDirectoryPath).File(sourceFilePath).File(homeFilePath).Build()
	defer cleanup(homeBackupFilePath)

	suite.command.SetArgs([]string{"--source-directory", sourceDirectoryPath, "--conflict-resolution", "backup"})
	suite.command.Execute()

	snaps.MatchSnapshot(suite.T(), suite.logs.String())
	homeFileStat, _ := os.Lstat(homeFilePath)
	assert.Equal(suite.T(), os.ModeSymlink, homeFileStat.Mode()&os.ModeSymlink)
	_, err := os.Lstat(homeBackupFilePath)
	assert.Nil(suite.T(), err)
}

func (suite *RootCommandSuite) TestShouldOverrideFileWhenFileExistsAndOverrideIsPassedToConflictResolutionOption() {
	suite.BeforeTest()

	sourceDirectoryPath := "/tmp/source/"
	sourceFilePath := "/tmp/source/foo.txt"
	homeFilePath := filepath.Join(suite.home, "foo.txt")
	cleanup := suite.fileSystemBuilder.Directory(sourceDirectoryPath).File(sourceFilePath).File(homeFilePath).Build()
	defer cleanup()

	suite.command.SetArgs([]string{"--source-directory", sourceDirectoryPath, "--conflict-resolution", "override"})
	suite.command.Execute()

	snaps.MatchSnapshot(suite.T(), suite.logs.String())
	homeFileStat, _ := os.Lstat(homeFilePath)
	assert.Equal(suite.T(), os.ModeSymlink, homeFileStat.Mode()&os.ModeSymlink)
}

func TestRootCommandSuite(t *testing.T) {
	suite.Run(t, &RootCommandSuite{})
}
