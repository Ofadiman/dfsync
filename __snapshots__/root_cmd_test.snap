
[TestRootCommandSuite/TestShouldExitEarlyWhenConflictResolutionFlagIsInvalid - 1]
DEBU command has been called with the following flags: --source-directory=/home/golang/dfsync, --dry=false, --conflict-resolution=bar
ERRO invalid value passed to --conflict-resolution flag, received bar

---

[TestRootCommandSuite/TestShouldExitEarlyWhenPathPassedToSourceDirectoryOptionDoesNotExist - 1]
DEBU command has been called with the following flags: --source-directory=/foo/bar/buzz, --dry=false, --conflict-resolution=no-action
ERRO path passed to --source-directory option does not exist, received /foo/bar/buzz

---

[TestRootCommandSuite/TestShouldExitEarlyWhenPathPassedToSourceDirectoryOptionIsNotDirectory - 1]
DEBU command has been called with the following flags: --source-directory=/tmp/foo.txt, --dry=false, --conflict-resolution=no-action
ERRO path passed to --source-directory option is not directory, received /tmp/foo.txt

---

[TestRootCommandSuite/TestShouldExitEarlyWhenSourceDirectoryIsEmpty - 1]
DEBU command has been called with the following flags: --source-directory=/tmp/source/, --dry=false, --conflict-resolution=no-action
ERRO directory passed to --source-directory option is empty, received /tmp/source/

---

[TestRootCommandSuite/TestShouldDoNothingWhenFileExistsAndNoActionIsPassedToConflictResolutionOption - 1]
DEBU command has been called with the following flags: --source-directory=/tmp/source/, --dry=false, --conflict-resolution=no-action
DEBU home directory: /home/golang/
DEBU absolute path to source directory: /tmp/source
DEBU processing path: /tmp/source
DEBU visiting source directory, no action required
DEBU processing path: /tmp/source/foo.txt
WARN path /home/golang/foo.txt already exists, symlink will not be created because --conflict-resolution flag is set to no-action

---

[TestRootCommandSuite/TestShouldCreateBackupFilesWhenFileExistsAndBackupIsPassedToConflictResolutionOption - 1]
DEBU command has been called with the following flags: --source-directory=/tmp/source/, --dry=false, --conflict-resolution=backup
DEBU home directory: /home/golang/
DEBU absolute path to source directory: /tmp/source
DEBU processing path: /tmp/source
DEBU visiting source directory, no action required
DEBU processing path: /tmp/source/foo.txt
WARN path /home/golang/foo.txt already exists, the file will be backed up and symlink will be created because --conflict-resolution flag is set to backup
INFO file has been successfully backup up under /home/golang/foo.txt.bak
INFO symlink from "/tmp/source/foo.txt" to "/home/golang/foo.txt" created

---

[TestRootCommandSuite/TestShouldCreateBackupFileWhenFileExistsAndBackupIsPassedToConflictResolutionOption - 1]
DEBU command has been called with the following flags: --source-directory=/tmp/source/, --dry=false, --conflict-resolution=backup
DEBU home directory: /home/golang/
DEBU absolute path to source directory: /tmp/source
DEBU processing path: /tmp/source
DEBU visiting source directory, no action required
DEBU processing path: /tmp/source/foo.txt
WARN path /home/golang/foo.txt already exists, the file will be backed up and symlink will be created because --conflict-resolution flag is set to backup
INFO file has been successfully backup up under /home/golang/foo.txt.bak
INFO symlink from "/tmp/source/foo.txt" to "/home/golang/foo.txt" created

---

[TestRootCommandSuite/TestShouldOverrideFileWhenFileExistsAndOverrideIsPassedToConflictResolutionOption - 1]
DEBU command has been called with the following flags: --source-directory=/tmp/source/, --dry=false, --conflict-resolution=override
DEBU home directory: /home/golang/
DEBU absolute path to source directory: /tmp/source
DEBU processing path: /tmp/source
DEBU visiting source directory, no action required
DEBU processing path: /tmp/source/foo.txt
WARN path /home/golang/foo.txt already exists, the file will be deleted and symlink will be created because --conflict-resolution flag is set to override
INFO symlink from "/tmp/source/foo.txt" to "/home/golang/foo.txt" created

---
